zenodo/modules/records/views.py

```diff
@@ -354,6 +354,31 @@ def record_from_pid(recid):
         return {}
 
 
+@blueprint.app_template_filter('sort_custom_keywords')
+def sort_custom_keywords(keywords):
+    custom_keywords_config = current_app.config.get(
+            'ZENODO_CUSTOM_METADATA_VOCABULARIES')
+
+    sortable_keywords = []
+    unsorted_keywords = []
+
+    for keyword in keywords:
+        (key, value) = keyword
+        (category, index) = key.split(':')
+        if category in custom_keywords_config.keys():
+            attribute = custom_keywords_config[category]['attributes'][index]
+            if 'order' in attribute.keys():
+                modified_keyword = (keyword[0], keyword[1], attribute['order'])
+                sortable_keywords.append(modified_keyword)
+            else:
+                unsorted_keywords.append(keyword)
+
+    sortable_keywords.sort(key=lambda x: x[2])
+    sorted_keywords = list(map(lambda x: (x[0], x[1]), sortable_keywords))
+
+    return sorted_keywords + unsorted_keywords
```

Comment:
- Regarding the return value of this specific case I would follow the paradigm of writing functions that accept small/light arguments, but return rich results that then the consumer is responsible for filtering. Thus I would say we return a list of dictionaries (instead of tuples), that you can then further process in the Jinja template via [`groupby`](https://jinja.palletsprojects.com/en/2.10.x/templates/#groupby), [`sort`](https://jinja.palletsprojects.com/en/2.10.x/templates/#sort), etc.
- In this case, it's actually fine to concatenate lists like this (since these will never be too large)